### PR TITLE
MINIFICPP-268 Implement ManipulateArchive processor

### PR DESCRIPTION
**Pull request by calebj**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: extensions/libarchive/ManipulateArchive.cpp [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+        }
+    }
+
+    // Perform operation: COPY
+    if (operation == OPERATION_COPY) {
+        bool found = false;
+        FocusArchiveEntry::ArchiveEntryMetadata copy;
+
+        // Find item to copy
+        for (auto it = archiveMetadata.entryMetadata.begin(); it != archiveMetadata.entryMetadata.end();) {
```
Comment (phrocker) [inline extensions/libarchive/ManipulateArchive.cpp]: Can clean this up with the range based loop, but that's not something that's absolutely necessary.

File: extensions/libarchive/ManipulateArchive.cpp [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+    session->read(flowFile, &readCallback);
+
+    logger_->log_info("ManipulateArchive performing operation %s on %s", operation.c_str(), targetEntry.c_str());
+
+    // Perform operation: REMOVE
+    if (operation == OPERATION_REMOVE) {
+        for (auto it = archiveMetadata.entryMetadata.begin(); it != archiveMetadata.entryMetadata.end();) {
+            if ((*it).entryName == targetEntry) {
+                logger_->log_info("ManipulateArchive found entry %s for removal", targetEntry.c_str());
+                std::remove((*it).tmpFileName.c_str());
```
Comment (phrocker) [inline extensions/libarchive/ManipulateArchive.cpp]: In the event that volatile repositories are used in environments without persistent storage should we disable this suite of extensions? As I understand it, these operations can be performed in memory, right @minifirocks?

Should we put make an option to exclude this and other processors for those without persistent storage?

File: extensions/libarchive/ManipulateArchive.cpp [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+        touchEntry.entryType = AE_IFREG;
+        touchEntry.entrySize = 0;
+        touchEntry.entryMTime = time(nullptr);
+        touchEntry.entryMTimeNsec = 0;
+        touchEntry.entryUID = 0;
+        touchEntry.entryGID = 0;
+        touchEntry.entryPerm = 0777;
+
+        // Update metadata
+        if (after != "") {
```
Comment (phrocker) [inline extensions/libarchive/ManipulateArchive.cpp]: All of these loops are essentially the same thing, can we combine them into a function and reuse the code?

Comment (calebj) [reply extensions/libarchive/ManipulateArchive.cpp]: Indeed. That's the next step; the first was getting the other processor merged, and then get this building on the ever-so-picky Xcode.

File: extensions/libarchive/ManipulateArchive.cpp (hunk 2 above)
Comment (calebj) [reply extensions/libarchive/ManipulateArchive.cpp]: I'm leaning towards yes. There's no surefire way to determine the uncompressed size of input files.

File: CMakeLists.txt [hunk 4]
```diff
@@ -111,38 +111,38 @@ add_subdirectory(libminifi)
...
 
 ## Add http-curl extensions
-createExtension(DISABLE_CURL 
-				HTTP-CURL 
-				"HTTP CURL" 
-				"This enables RESTProtocol, InvokeHTTP, and the HTTPClient for Site to Site" 
-				"extensions/http-curl"
-				"${TEST_DIR}/curl-tests")
+createExtension(DISABLE_CURL
+                HTTP-CURL
```
Comment (phrocker) [inline CMakeLists.txt]: we need to add a skip to this so this doesn't happen when the linter runs. thanks

Comment (calebj) [reply CMakeLists.txt]: I don't follow. What do you need me to change?

**Comment by calebj** [comment]
Is this waiting on #181? I have another branch based on that code I can push this PR if needed. Otherwise, is there an estimate as to when this might be merged?

**Comment by phrocker** [comment]
@calebj yes, we're removing Boost dependencies.

**Comment by phrocker** [comment]
@calebj I can make the changes to boost if you're fine with it when I merge it, unless you are doing that now.

**Comment by calebj** [comment]
Should be ready now. Check to see if the exceptions and error output on misconfigurations fits what you want.

**Pull request closed without merging**